builtin/providers/cloudstack/resource_cloudstack_private_gateway.go

```diff
@@ -0,0 +1,173 @@
+package cloudstack
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/xanzy/go-cloudstack/cloudstack"
+)
+
+func resourceCloudStackPrivateGateway() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceCloudStackPrivateGatewayCreate,
+		Read:   resourceCloudStackPrivateGatewayRead,
+		Update: resourceCloudStackPrivateGatewayUpdate,
+		Delete: resourceCloudStackPrivateGatewayDelete,
+
+		Schema: map[string]*schema.Schema{
+			"gateway": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"ip_address": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"netmask": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"vlan": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"physical_network_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"network_offering": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"acl_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"vpc_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceCloudStackPrivateGatewayCreate(d *schema.ResourceData, meta interface{}) error {
+	cs := meta.(*cloudstack.CloudStackClient)
+
+	ipaddress := d.Get("ip_address").(string)
+	networkofferingid := d.Get("network_offering").(string)
+
+	// Create a new parameter struct
+	p := cs.VPC.NewCreatePrivateGatewayParams(
+		d.Get("gateway").(string),
+		ipaddress,
+		d.Get("netmask").(string),
+		d.Get("vlan").(string),
+		d.Get("vpc_id").(string),
+	)
+
+	// Retrieve the network_offering ID
+	if networkofferingid != "" {
+		networkofferingid, e := retrieveID(cs, "network_offering", networkofferingid)
+		if e != nil {
+			return e.Error()
+		}
+		p.SetNetworkofferingid(networkofferingid)
+	}
+
+	// Check if we want to associate an ACL
+	if aclid, ok := d.GetOk("acl_id"); ok && aclid.(string) != none {
```

Comment:
This test (against `none`) is no longer needed here as this part was refactored, remember?